### ISSUE TITLE
Bundaberg Regional Council

### ISSUE THREAD

**Issue by mlandauer** [planningalerts-scrapers/issues#964]
This issue has been **automatically** created by PlanningAlerts. Only close this issue once the authority is working again on PlanningAlerts.

**Comment by katska** [comment]
error

bundaberg: ERROR: Net::ReadTimeout
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/protocol.rb:181:in `rbuf_fill'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/protocol.rb:157:in `readuntil'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/protocol.rb:167:in `readline'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/http/response.rb:40:in `read_status_line'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/http/response.rb:29:in `read_new'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/http.rb:1494:in `block in transport_request'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/http.rb:1491:in `catch'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/http.rb:1491:in `transport_request'
/app/vendor/ruby-2.5.1/lib/ruby/2.5.0/net/http.rb:1464:in `request'
/app/vendor/bundle/ruby/2.5.0/gems/net-http-persistent-4.0.1/lib/net/http/persistent.rb:891:in `block in request'
/app/vendor/bundle/ruby/2.5.0/gems/net-http-persistent-4.0.1/lib/net/http/persistent.rb:606:in `connection_for'
/app/vendor/bundle/ruby/2.5.0/gems/net-http-persistent-4.0.1/lib/net/http/persistent.rb:885:in `request'
/app/vendor/bundle/ruby/2.5.0/gems/mechanize-2.7.7/lib/mechanize/http/agent.rb:280:in `fetch'
/app/vendor/bundle/ruby/2.5.0/gems/mechanize-2.7.7/lib/mechanize.rb:1323:in `post_form'
/app/vendor/bundle/ruby/2.5.0/gems/mechanize-2.7.7/lib/mechanize.rb:536:in `post'
/app/lib/masterview_scraper/get_applications_api.rb:47:in `scrape_page'
/app/lib/masterview_scraper/get_applications_api.rb:18:in `scrape'
/app/lib/masterview_scraper.rb:97:in `scrape_api_period'
/app/lib/masterview_scraper.rb:47:in `scrape_period'
/app/lib/masterview_scraper.rb:19:in `scrape'
scraper.rb:16:in `block in scrape'
scraper.rb:12:in `each'
scraper.rb:12:in `scrape'
scraper.rb:33:in `<main>'

**Comment by katska** [comment]
In a browser, I can reach https://da.bundaberg.qld.gov.au/ and https://da.bundaberg.qld.gov.au/Home/Disclaimer and 
https://da.bundaberg.qld.gov.au/Application/ApplicationDetails/522.2019.00000145.001
listed in the scraper's yml file ok, but https://da.bundaberg.qld.gov.au/Application/GetApplications no longer exists - its a 404

Applications on exhibition now at:
https://da.bundaberg.qld.gov.au/Application/AdvancedSearchResult?RemoveUndeterminedApplications=False&ShowOutstandingApplications=False&ShowExhibitedApplications=True&IncludeDocuments=False in a browser just fine 
and a recent application listed is at https://da.bundaberg.qld.gov.au/Application/ApplicationDetails/522.2022.00000364.001/
which is suspiciously similar. 
@mlandauer does reading any of this give you a clue what's going on here?

**Comment by ianheggie-oaf** [comment]
Works with my dev system, but was 403 (refused) on morph.io - retrying with https://github.com/ianheggie-oaf/multiple_masterview/commit/35fb6f9436b9bee71a8a472f9441b91cc5eb04b6

**Comment by katska** [comment]
oh so that seems likely blocked by council. like #913 - @ianheggie-oaf is there anything else you can do to check? 
I added the label anti-scraping technology as a pointer to look at others. You might want to look in the code of other scrapers (I've seen reference in the technology_one multi scraper for example, that the scraper is run with Aussie proxy for example, but I don't know what the difference is in error returned or other behavior in that case, sorry).

**Comment by ianheggie** [comment]
@katska I appreciate you following up with the council on fixing the access

FYI @mlandauer -I have added a flag so it wont complain, but it will keep checking - it will throw and error when they start working and need to come off the list: Extract from CHANGELOG.md in [My Repo](https://github.com/ianheggie-oaf/multiple_masterview) - I have some final testing before releasing it to you to install (or you can enable fork and pull requests for me (ianheggie-oaf) from the private repo)

* Use MORPH_EXPECT_BAD=albury,bega_valley,bundaberg,port_stephens 
  as they have anti-scraping errors!
  * albury: ERROR: Failed to open TCP connection to eservice.alburycity.nsw.gov.au:443 (execution expired) after 45.1 seconds
  * bega_valley: ERROR: Failed to open TCP connection to datracker.begavalley.nsw.gov.au:443 (execution expired) after 45.0 seconds
  * bundaberg: ERROR: 403 => Net::HTTPForbidden for https://da.bundaberg.qld.gov.au/ -- unhandled response after 0.2 seconds
  * port_stephens: ERROR: Failed to open TCP connection to datracker.portstephens.nsw.gov.au:443 (execution expired) after 45.0 seconds

**Comment by ianheggie** [comment]
@katska Actually please assign to yourself for follow up, or de-assign as I can't do anything more with this

I have provided and recommended setting `MORPH_EXPECT_BAD=albury,bogan,bega_valley,bundaberg,griffith,port_stephens,port_macquarie_hastings` so the scraper will silently keep trying these sites and then alert us when one or more start working!

**Comment by ianheggie** [comment]
@katska (and @mlandauer ) BTW - In answer to "is there anything else you can do to check?" - yes, as per the report below, when I cleared the test cache and forced it to get a fresh cache of each councils site it WAS able to successfully extract records from most sites (including this one) WHEN run on my system (except for bogan and griffith). Running my code on morph it was able to also get records from the same sites apart from port_macquarie_hastings.

```
ianh@eli:.../completed/multiple_masterview (master)$ bin/rake morph:status
Authority               Week Month Population Test Morph Ago Err% Warning
----------------------- ---- ----- ---------- ---- ----- --- ---- ----------------------------------------
albury                     0     0     56,093   96    67  16  100 Last application received 6 months ago.
ballina                    0     0     46,296   99   128   1   17 Last application received 6 months ago.
bega_valley                0     0     35,942   19    15  16  100 Last application received 6 months ago.
bogan                      0     0      2,467    0     -   -  100 Last application received 10 months ago.
broken_hill                0     0     17,588    3     5   1   17 Last application received 6 months ago.
bundaberg                  0     0     99,215   72    49  16  100 Last application received 8 months ago.
byron                      0     0     36,116   43    82   1   17 Last application received 6 months ago.
cessnock                   0     0     63,632  205   280   1   33 Last application received 6 months ago.
dubbo                      0     0     54,922  132   169   1   17 Last application received 6 months ago.
fairfield                  0     0    208,475   37    46   1   33 Last application received 6 months ago.
griffith                   0     0     27,086    0     -   -  100 Last application received 6 months ago.
gympie                     0     0     53,242   20    29   1   17 Last application received 6 months ago.
lismore                    0     0     44,334   50    59   1   17 Last application received 6 months ago.
maranoa                    0     0     12,825   19    24   1   17 Last application received 6 months ago.
muswellbrook               0     0     16,357   18    24   1   17 Last application received 6 months ago.
port_macquarie_hastings    0     0     86,762  296     -   -  100 Last application received 6 months ago.
port_stephens              0     0     75,276   85    64  16  100 Last application received 6 months ago.
shoalhaven                 0     0    108,531  122   159   1   17 Last application received 6 months ago.
singleton                  0     0     24,577   46    67   1   33 Last application received 6 months ago.
strathfield                0     0     45,593   24    36   1   17 Last application received 6 months ago.
upper_hunter               0     0     14,229   49    67   1   17 Last application received 6 months ago.

KEY:
Week, Month, Population, Warning - from Morph.io Live Status Page
Test - local spec/expected/AUTHORITY.yml record count
Morph - records collected by test of this repo on Morph.io
Err% - the percentage of runs that had complete or partial (record) errors
Ago - how many days ago the last record was added by this repo on Morph.io

```

**Comment by katska** [comment]
Ok @ianheggie-oaf from the output above I'm seeing
*  there's a 403, 
* that the scraper works 'on your system' which I take to mean locally when you clear the cache
* It errors when run on morph
* There isn't any more information from the errors or your investigations that sheds further light on what's happening at Bundaberg 
* You don't explicitly say this but the error and behaviour look typical of Antiscraping technology, so applying that label.

I'm going to open another issue in this repo so I can ask questions that apply to multiple masterview using councils, rather than mentioning what's happening across different councils in another council's issue, as I find that confusing.

**Comment by ianheggie-oaf** [comment]
Confirmed when I run tests with the standard scraper and then MY Australian proxy, it starts working => there is "anti scraping tech" which is banning scraping sites after some period of time. I recommend we consider refreshing the proxy twice a week to get a fresh IP#.

See
- https://github.com/planningalerts-scrapers/issues/issues/1071

With standard proxy:
```
Scraping Summary:
Authority                OK    Bad Exception
-------------------- ------ ------ --------------------------------------------------
albury                    0      0 Net::OpenTimeout - execution expired
ballina                   0      0 Net::OpenTimeout - execution expired
bega_valley               0      0 Net::OpenTimeout - execution expired
broken_hill               0      0 Net::OpenTimeout - execution expired
bundaberg                 0      0 Net::OpenTimeout - execution expired
burwood                   0      0 ArgumentError - unknown keyword: australian_proxy
byron                     0      0 Net::OpenTimeout - execution expired
cessnock                  0      0 Net::OpenTimeout - execution expired
dubbo                     0      0 Net::OpenTimeout - execution expired
fairfield                 0      0 Net::OpenTimeout - execution expired
gympie                    0      0 Net::OpenTimeout - execution expired
lismore                   0      0 Net::OpenTimeout - execution expired
maranoa                   0      0 Net::OpenTimeout - execution expired
muswellbrook              0      0 Net::OpenTimeout - execution expired
port_stephens             0      0 Net::OpenTimeout - execution expired
strathfield               0      0 Net::OpenTimeout - execution expired
upper_hunter              0      0 Net::OpenTimeout - execution expired
```

VS With my proxy on https://morph.io/ianheggie-oaf/multiple_masterview-prs
```
Scraping Summary:
Authority                OK    Bad Exception
-------------------- ------ ------ --------------------------------------------------
albury                  194      0 -
ballina                 159      0 -
bega_valley              26      0 -
broken_hill               8      0 -
bundaberg                70      0 -
burwood                  49      0 -
byron                    88      0 -
cessnock                  0      0 -
dubbo                   151      0 -
fairfield                36      0 -
gympie                   37      0 -
lismore                  16      0 RuntimeError - Unexpected field in: Determination Type: on https://tra
maranoa                  25      0 Net::ReadTimeout - Net::ReadTimeout
muswellbrook             21      0 -
port_stephens           132      0 -
strathfield              50      0 -
upper_hunter             47      0 -
```

**Comment by ianheggie-oaf** [comment]
@simonzippy - Confirming production works when rotating proxy IP by changing production to using my rotating IP proxy - this validates the need for:
- https://github.com/planningalerts-scrapers/issues/issues/1071
